# Math 134C Week 3

In [2]:
import math
# r is the risk-free rate, F is a cash flow, and dT is the time that we get each cash flow
def PV(F, dT, r):
    pv = 0
    if len(dT) != len(F):
        print("Input Error")
        return 0
    for n in range(0, len(F)):
        pv += F[n] * math.exp(-r * dT[n])
    return pv
        
# C_t is the price of call option, S_t is the current stock price, K is the strike, r is the risk-free rate, dt is time to expiration. D is the dividend of the stock, and dT is the time at which we receive the dividend. The delta is for the continuous dividend paying model. The function will return the price of the corresponding put option. 
def put_given_call(C_t, S_t, K, r, dt, D= None, dT = None, delta = 0):
    if D is None: D = []
    if dT is None: dT = []
    return C_t - (S_t - PV(D, dT, r)) * math.exp(-delta * dt) + K * math.exp(-r * dt)

def call_given_put(P_t, S_t, K, r, dt, D = None, dT = None, delta = 0):
    if D is None: D = []
    if dT is None: dT = []
    return P_t + (S_t - PV(D, dT, r)) * math.exp(-delta * dt) - K * math.exp(-r * dt)

def one_step_binomial(r, dt, u, d, Vu, Vd, delta = 0):
    q = (math.exp((r - delta) * dt) - d)/(u-d)
    if not (1 > q > 0): 
        print("Arbitrage opportunity exists")
        return -1
    return math.exp(-r * dt)*(q*Vu+(1-q)*Vd)    

def Delta(r, dt, u, d, Vu, Vd, S, delta = 0):
    return math.exp(-delta * dt)*(Vu-Vd)/(S*(u-d))

def B(r, dt, u, d, Vu, Vd):
    return math.exp(-r * dt)*(u * Vd - d * Vu)/(u-d)

# One Step Binomial Model for Option Pricing
Let $V_t$ to be the value of an investment instrument at time $t$ whose payoff at time $T$ only depends on the stock price. (For example, V could be a call option, put option, future, or other products). Let $S_t$ be the price of the stock at time $t$, $\delta$ be the continuous dividend yield, and $u, d$ be the number such that the stock will either go up to $Su$ or go down to $Sd$ at time $T$. Let $V_T^u$ be the value of the investment instrument at time $T$ when the stock goes to $Su$, and $V_T^d$ be the value when the stock goes to $Sd$ (If $V$ is a call option with strike $K$, then $V_T^u = (Su-K)^+, V_T^d = (Sd-K)^+$). Assume that the risk-free rate is $r$, and denote $T-t$ as $\Delta t$. Then we define the risk-neutral probability measure (also called price measure) 
$$q = \frac{e^{(r-\delta) \Delta t} - d}{u-d}$$
Then the price of $V$ at time $t$ is given by 
$$V_t = e^{-r\Delta t}(qV_T^u + (1-q)V_T^d)$$

**Self-financing Replicating Portfolio:**
We will construct a portfolio 
$$\Pi_t = \Delta S_t + B_t$$
consisting of $\Delta$ share of the stock, and $B_t$ amount of money in the bank. We wish to use this portfolio to replicate the payoff of a financial instrument $V_t$. Then we must have 
$$\Pi_T^u = \Delta S_tue^{\delta \Delta t} + B_te^{r \Delta t} = V_T^u$$
$$\Pi_T^d = \Delta S_tde^{\delta \Delta t} + B_te^{r \Delta t} = V_T^d$$
where $\delta$ is the continuous dividend. ($\Delta$ share of the stock grow to $\Delta e^{\delta \Delta t}$ share)
By solving the above equations, we get 
$$\Delta = e^{-\delta \Delta t}\frac{V_T^u-V_T^d}{S_t(u-d)}$$
$$B_t = e^{-r \Delta t} \frac{u V_T^d - dV_T^u}{u-d}$$
$$V_t = \Pi_t = e^{-\delta \Delta t}\frac{V_T^u-V_T^d}{u-d} +  e^{-r \Delta t} \frac{u V_T^d - dV_T^u}{u-d} = e^{-r\Delta t}(qV_T^u + (1-q)V_T^d)$$

**Quiz 2.1**:
You have 40 years until retirement and currently have all your wealth invested in the stock market.  

Very roughly speaking, an extra 1% return per year in the stock market has what effect on your wealth at retirement?

In [2]:
print(f"Our wealth will grow by {(1+0.01)**40 - 1:.2f}")

Our wealth will grow by 0.49


**Quiz 2.2:**
Using historical daily closing prices, You determine the daily sample standard deviation for Apple (ticker:  AAPL) is 1.58%.

What is the annual sample standard deviation (volatility) for Apple?

In [7]:
print(f"The annual volatility is: {1.58 * math.sqrt(252):.2f}%")

The annual volatility is: 25.08%


**Quiz 2.3:**
You are given the following information for a one-period binomial tree

Each period is one year
The current price of the stock S = 100
The stock pays dividends at a rate proportional to its price with the dividend yield d = 2%
The continuously compounded risk-free interest rate is r
u = 1.3, where Su is the stock price if the stock price goes up
d = 0.7, where Sd is is the stock price if the stock price goes down
The current premium of a one-year at-the-money put option is 12.82
Immediately after purchasing the put option, the stock price drops to 95.

Assuming that u, d, r, and the strike price of the put option remain unchanged, recalculate the premium of the put option.

In [8]:
print(f"The price for the new put option is {12.82/30 * (100-95*0.7):.2f}")

The price for the new put option is 14.32


**Quiz 2.4:**
Recall that a straddle position is a long call option plus a long put option with the same strike price and same expiration date.

For a one-year straddle on a nondividend-paying stock, you are given:

The straddle can only be exercised at the end of one year.
The payoff of the straddle is the absolute value of the difference between the strike price and the stock price at expiration date.
The stock currently sells for 60.00.
The continuously compounded risk-free interest rate is 8%.
In one year, the stock will either sell for 70.00 or 45.00.
The options have a strike price of 50.00.
Using a one-period binomial options pricing model, calculate the current price of the straddle.

In [9]:
print(f"The current price of the straddle under the one-step binomial model is {one_step_binomial(r = 0.08, dt = 1, u = 70/60, d = 45/60, Vu = 20, Vd = 5):.2f}")

The current price of the straddle under the one-step binomial model is 15.69


**Quiz 2.5:**
Let $S_t$ be the time-t price of the stock.

For a one-period binomial model, you are given:

Each period is 3 years.
The current price for a stock $S_0 = 100$
The stock pays dividends continuously at a rate proportional to its price. The dividend yield is 3%
The stock price at the end of three years is $S_3$ 
$S_3 = S_0u$ if the stock price goes up, where $u = 1.30$
$S_3 = S_0d$ if the stock price goes down, where $d = 0.85$ 
The continuously compounded risk-free annual interest rate is 6%.
Consider a derivative that at the end of three years pays $(S_3-S_0)^4$

Calculate the number of shares of stocks an investor must purchase at time 0 in order to replicate the derivative.

In [15]:
delta = Delta(r = 0.06, dt = 3, u = 1.3, d = 0.85, Vu = 30**4, Vd = 15**4, S=100, delta=0.03)
print(f"The number of shares of stocks we should buy is {delta:.0f}")

The number of shares of stocks we should buy is 15423


**Quiz 2.6:**
For a two-period binomial model for stock prices, you are given:

Each period is one year.
The current price of the stock is 100.
The stock pays dividends continuously at a rate proportional to its price. The dividend yield is 3\%.
The continuously compounded risk-free interest rate is 5\%.
u = 1.1, where S goes from S to Su during each period if the stock price goes up
d = 0.9, where S goes from S to Sd during each period if the stock price goes down
A two-year European put option on the stock has a strike price K, where K < 120

Calculate the value of K such that the current price of the put option is 11.84.

In [12]:
q = (math.exp(0.05- 0.03)-0.9)/(1.1-0.9)
print(f"The price measure q is given by {q:.2f}")
K = (11.84 + math.exp(-2 * 0.05) * ((2*q-2*q**2)*99 + (1-q)**2 * 81))/(math.exp(-2 * 0.05) * (2*q - 2*q**2+(1-q)**2))
print(f"The strike should be {K:.2f}")

The price measure q is given by 0.60
The strike should be 115.00


**Quiz 2.7:**
A European call option on a stock is priced using a two-period binomial model. The call price tree is as follows:
![binomial tree](../img/134C_Week3_image.png "Title")
Calculate the current price of the call option $C_0$

In [5]:
q = (-0.474/13.235 * 1.092)/(0.474/13.235 * (29.149-1.092) - 1.092)
d = 0.474/(q * 1.092)
C = d* (q * 13.232 + (1-q) * 0.474)
print(f"The price measure q is {q:.2f}")
print(f"The discount factor per period is {d:.2f}")
print(f"The price of this call option is {C:.2f}")

The price measure q is 0.45
The discount factor per period is 0.97
The price of this call option is 6.00


**Quiz 2.8:**
Consider the following one-period binomial model for a stock that pays dividends continuously at a rate proportional to its price:


You are given the following information:
- The stock price will be either 450 or 325 in the next period
- Each period length is 3 months.
- The continuously compounded risk-free interest rate is 10%.
- The continuous annual dividend yield on the stock is $\delta$
An investor wishes to replicate a 3-month European call option on the stock with a strike price of 350.

Determine the amount in dollars the investor must lend at the risk-free rate.

**Quiz 2.9:**
For a six-month 40-50 strangle on a stock paying dividends continuously at a annual rate of 2%, you are given:

- The strangle can only be exercised at the end of six months.
- The payoff of the strangle is given $\max(0, S-50)+\max(0, 40-S)$
- The stock currently sells for 45.00.
- In six months, the stock will either sell for 58.50 or 36.00.
- The continuously compounded annual risk-free interest rate is 7\%.
Calculate the current price of the 40-50 strangle.

In [13]:
print(f"The current price for the strangle is {one_step_binomial(r = 0.07, dt = 0.5, u = 58.5/45, d = 36/45, Vu = 8.5, Vd = 4, delta = 0.02):.2f}")

The current price for the strangle is 5.82


**Quiz 2.10:**
A non-dividend paying stock is modeled by the binomial tree where its price is 10 at $t = 0$, and it can be either 18 or 4 at $t = 1$.

A European call option on the stock above expires at t=1 year with strike price K = 12.

Calculate the number of shares of stock in the replicating portfolio for this option.

# American Option
It is easy to check at each note if the early exercise is optimal, the binomial method is well-suited to valuing American options. For example, the value of the call option at each node is given by the maximum of $(S-K)^+$, which is the value if we choose to exercise at this point, and its price if it is not exercised.

**Exercise 10.10:**
Let S = 100, K = 95, $\sigma$ = 30\%, r = 8\%, T = 1, and $\delta$ = 0. Let u = 1.3, d = 0.8, and n = 2. Construct the binomial tree for an American put option. At each node provide the premium, $\Delta$, and B.


In [9]:
import math
q = (math.exp(0.04) - 0.8)/(1.3 - 0.8)
P = math.exp(-0.04) * (q * 0 + (1-q) * 31)
P0 = math.exp(-0.04) * (q * 0 + (1-q) * P)
print(f"The price measure q is given by {q:.2f}")
print(f"The value of the put option at t = 0.5 if we choose not to exercise is {P:.2f}")
print(f"The value of the put option at t = 0 if we choose not to exercise is {P0:.2f}")

The price measure q is given by 0.48
The value of the put option at t = 0.5 if we choose not to exercise is 15.44
The value of the put option at t = 0 if we choose not to exercise is 7.69


**Exercise 11.1**:
Consider a one-period binomial model with h = 1, where S = 100, r = 0, $\sigma$ = 30%, and $\delta$ = 0.08. Compute American call option prices for K = 70, 80, 90, and 100.
a. At which strike(s) does early exercise occur?
b. Use put-call parity to explain why early exercise does not occur at the higher strikes.
c. Use put-call parity to explain why early exercise is sure to occur for all lower strikes than that in your answer to (a).

In [15]:
u = math.exp(0.08 + 0.3)
d = math.exp(0.08 - 0.3)
print(f"The u is {u:.2f}, and d is {d:.2f}")
q = (math.exp(-0.08) - d)/(u-d)
print(f"The price measure q is {q:.2f}")
K = 70
C = q * max(0,  100 * u- K ) + (1-q) * max(0,  100 * d - K)
print(f"The price of the call option if we choose not to exercise early is {C:.2f}")
print(f"The value of the call option if we choose to exercise early is {max(100-K, 0)}")

The u is 1.46, and d is 0.80
The price measure q is 0.18
The price of the call option if we choose not to exercise early is 22.31
The value of the call option if we choose to exercise early is 30


**Exercise 11.4**
Consider a one-period binomial model with $h=1$, where $S = 100$, $r = 0.08$, $\sigma = 0.3$, $\delta = 0$. Compute American put option prices for $K = 100, 110, 120$ and $130$. \
a) At which strikes does the early exercise occur?\
b) Use put-call parity to explain why early exercise does not occur at the other strikes \
c) Use put-call parity to explain why early exercise is sure to occur for all strikes greater than that in your answer to a)

In [42]:
u = math.exp(0.08 + 0.3)
d = math.exp(0.08 - 0.3)
print("a) ")
for strike in [100, 110, 120, 130]:
    price_no_exercise = one_step_binomial(0.08, 1, u, d, max(strike - 100 * u,0), max(strike - 100 * d, 0))
    price_exercise_early = max(strike - 100, 0)
    if price_exercise_early < price_no_exercise:
        print(f"Strike {strike}; European option price: {price_no_exercise:.2f}; Early exercise value: {price_exercise_early}; American option price: {price_no_exercise:.2f}; No early exercise\n") 
    else:
        print(f"Strike {strike}; European option price: {price_no_exercise:.2f}; Early exercise value: {price_exercise_early}; American option price: {price_exercise_early:.2f}; Early exercise\n") 

a) 
Strike 100; European option price: 10.47; Early exercise value: 0; American option price: 10.47; No early exercise

Strike 110; European option price: 15.77; Early exercise value: 10; American option price: 15.77; No early exercise

Strike 120; European option price: 21.08; Early exercise value: 20; American option price: 21.08; No early exercise

Strike 130; European option price: 26.38; Early exercise value: 30; American option price: 30.00; Early exercise


b) By the put-call parity, we have 
$$P = C-S+Ke^{-0.08}$$
We will exercise early if and only if $(K-S)^+ \geq P$. Since $P$ is always non-negative, the statement is equivalent to $K-S \geq P$, and we have
$$K-S\geq C-S+Ke^{-0.08}$$
Hence, we choose to exercise early if the following equality holds
$$K(1-e^{-0.08})\geq C$$

We know that the early exercise does not occur at $K = 120$, and the above inequality does not hold at this $K$. For smaller values of $K$, $C$ will be larger, and thus the above inequality cannot hold. Hence, we will not choose to exercise early at strike lower than 120. 

c) The early exercise will occur at strike greater than 130. This is because the above inequality holds for 130, and for larger $K$, the $C$ will be smaller, and thus the inequality will always hold. Hence, we will choose to exercise early at strike greater than 130, 

**Exercise 11.5**
Repeat the above exercise with $\delta = 0.08$. What is the lowest strike price at which early exercise will occur? What condition related to put-call parity is satisfied at this strike price?

In [40]:
u = math.exp(0.3)
d = math.exp(-0.3)
print("a) ")
for strike in [100, 110, 120, 130]:
    price_no_exercise = one_step_binomial(0.08, 1, u, d, max(strike - 100 * u,0), max(strike - 100 * d, 0), 0.08)
    price_exercise_early = max(strike - 100, 0)
    if price_exercise_early < price_no_exercise:
        print(f"Strike {strike}; European option price: {price_no_exercise:.2f}; Early exercise value: {price_exercise_early}; American option price: {price_no_exercise:.2f}; No early exercise\n") 
    else:
        print(f"Strike {strike}; European option price: {price_no_exercise:.2f}; Early exercise value: {price_exercise_early}; American option price: {price_exercise_early:.2f}; Early exercise\n") 

a) 
Strike 100; European option price: 13.74; Early exercise value: 0; American option price: 13.74; No early exercise

Strike 110; European option price: 19.05; Early exercise value: 10; American option price: 19.05; No early exercise

Strike 120; European option price: 24.35; Early exercise value: 20; American option price: 24.35; No early exercise

Strike 130; European option price: 29.65; Early exercise value: 30; American option price: 30.00; Early exercise


b) In the computation above, the lowest strike that the early exercise happen is $K = 130$. 

The early exercise will happen if and only if
$$K-S\geq C - Se^{-0.08} + Ke^{-0.08}$$
which is equivalent to 
$$(1-e^{-0.08})K\geq C + 100 \cdot (1-e^{-0.08})$$

**Exercise 11.6**
Repeat the exercise 11.4 with $r = 0$, $\delta = 0.08$. What is the lowest strike price (if there is one) at which early exercise will occur? If exercise never occurs, explain why not. 

In [41]:
u = math.exp(-0.08 + 0.3)
d = math.exp(-0.08 -0.3)
print("a) ")
for strike in [100, 110, 120, 130]:
    price_no_exercise = one_step_binomial(0, 1, u, d, max(strike - 100 * u,0), max(strike - 100 * d, 0), 0.08)
    price_exercise_early = max(strike - 100, 0)
    if price_exercise_early < price_no_exercise:
        print(f"Strike {strike}; European option price: {price_no_exercise:.2f}; Early exercise value: {price_exercise_early}; American option price: {price_no_exercise:.2f}; No early exercise\n") 
    else:
        print(f"Strike {strike}; European option price: {price_no_exercise:.2f}; Early exercise value: {price_exercise_early}; American option price: {price_exercise_early:.2f}; Early exercise\n") 

a) 
Strike 100; European option price: 18.16; Early exercise value: 0; American option price: 18.16; No early exercise

Strike 110; European option price: 23.90; Early exercise value: 10; American option price: 23.90; No early exercise

Strike 120; European option price: 29.65; Early exercise value: 20; American option price: 29.65; No early exercise

Strike 130; European option price: 37.69; Early exercise value: 30; American option price: 37.69; No early exercise


The early exercise will never occur. This is because that early exercise will occur if and only if the inequality holds
$$K-S\geq C - Se^{-0.08} + K$$
However, the above inequality is equivalent to 
$$C\leq 100 * (e^{-0.08}-1) = -7.69$$
This can never happen, since the price of a call option should be non-negative by no-arbitrage principle.